### PR TITLE
Fix Maintain Sorted Mechanism 

### PR DESCRIPTION
**Pull request by fustilio**
Maintain sorted now selects the most recently added and edited.

Still retains structure to reselect previous parcel card

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/AddCommand.java [hunk 1]
```diff
@@ -61,10 +61,7 @@ public CommandResult executeUndoableCommand() throws CommandException {
         try {
             model.addParcel(toAdd);
             model.maintainSorted();
-            if (model.hasSelected()) {
```
Comment (Kennard123661) [inline src/main/java/seedu/address/logic/commands/AddCommand.java]: try to make maintainSorted called in model rather than in the command itself. Like perhaps put it in the internal logic of addParcel. Then model.reselect(toAdd) can also be in addParcel().

**Review by Kennard123661 — CHANGES_REQUESTED** [review]
I think by putting it in the command, you are violating one of the principles

**Comment by ktnc985** [comment]
1. Sort by Delivery date
2. Sort by Name
3. Sort by Tracking number

File: src/main/java/seedu/address/logic/commands/AddCommand.java [hunk 2]
```diff
@@ -59,12 +59,7 @@ public AddCommand(ReadOnlyParcel parcel) {
     public CommandResult executeUndoableCommand() throws CommandException {
         requireNonNull(model);
         try {
-            model.addParcel(toAdd);
-            model.maintainSorted();
-            if (model.hasSelected()) {
-                model.reselect(toAdd);
-            }
-
+            model.addParcelCommand(toAdd);
```
Comment (Kennard123661) [inline src/main/java/seedu/address/logic/commands/AddCommand.java:62]: Looks good

File: src/main/java/seedu/address/logic/commands/ClearCommand.java [hunk 3]
```diff
@@ -17,7 +17,6 @@
     public CommandResult executeUndoableCommand() {
         requireNonNull(model);
         model.resetData(new AddressBook());
-        model.unselect();
         return new CommandResult(MESSAGE_SUCCESS);
```
Comment (Kennard123661) [inline src/main/java/seedu/address/logic/commands/ClearCommand.java:20]: Good

File: src/main/java/seedu/address/logic/commands/EditCommand.java [hunk 4]
```diff
@@ -86,18 +86,15 @@ public CommandResult executeUndoableCommand() throws CommandException {
         Parcel editedParcel = createEditedParcel(parcelToEdit, editParcelDescriptor);
 
         try {
-            model.updateParcel(parcelToEdit, editedParcel);
-            model.maintainSorted();
-            if (model.hasSelected()) {
-                model.reselect(editedParcel);
-            }
-
+            model.editParcelCommand(parcelToEdit, editedParcel);
```
Comment (Kennard123661) [inline src/main/java/seedu/address/logic/commands/EditCommand.java:89]: Good

File: src/main/java/seedu/address/logic/commands/FindCommand.java [hunk 5]
```diff
@@ -24,7 +24,6 @@ public FindCommand(NameContainsKeywordsPredicate predicate) {
     @Override
     public CommandResult execute() {
         model.updateFilteredParcelList(predicate);
-        model.unselect();
```
Comment (Kennard123661) [inline src/main/java/seedu/address/logic/commands/FindCommand.java:27]: Good

File: src/main/java/seedu/address/logic/commands/RedoCommand.java [hunk 6]
```diff
@@ -25,7 +25,6 @@ public CommandResult execute() throws CommandException {
         }
 
         undoRedoStack.popRedo().redo();
-        model.maintainSorted();
```
Comment (Kennard123661) [inline src/main/java/seedu/address/logic/commands/RedoCommand.java:28]: Good

File: src/main/java/seedu/address/logic/commands/SelectCommand.java [hunk 7]
```diff
@@ -39,8 +39,6 @@ public CommandResult execute() throws CommandException {
         }
 
         EventsCenter.getInstance().post(new JumpToListRequestEvent(targetIndex));
-        model.setPrevIndex(targetIndex);
```
Comment (Kennard123661) [inline src/main/java/seedu/address/logic/commands/SelectCommand.java:42]: Good

File: src/main/java/seedu/address/logic/commands/UndoCommand.java [hunk 8]
```diff
@@ -25,7 +25,6 @@ public CommandResult execute() throws CommandException {
         }
 
         undoRedoStack.popUndo().undo();
-        model.maintainSorted();
```
Comment (Kennard123661) [inline src/main/java/seedu/address/logic/commands/UndoCommand.java:28]: Good

File: src/main/java/seedu/address/model/ModelManager.java [hunk 9]
```diff
@@ -57,6 +60,7 @@ public ModelManager(ReadOnlyAddressBook addressBook, UserPrefs userPrefs) {
         filteredParcels = new FilteredList<>(this.addressBook.getParcelList());
         updatedDeliveredAndUndeliveredList();
         activeFilteredList = filteredUndeliveredParcels;
+        ModelListener sl = new ModelListener(this);
```
Comment (Kennard123661) [inline src/main/java/seedu/address/model/ModelManager.java]: pick a better name for ModeListener

File: src/main/java/seedu/address/logic/commands/UndoableCommand.java [hunk 10]
```diff
@@ -33,6 +35,12 @@ protected final void undo() {
         requireAllNonNull(model, previousAddressBook);
         model.resetData(previousAddressBook);
         model.updateFilteredParcelList(PREDICATE_SHOW_ALL_PARCELS);
+        model.setActiveList(!previousActiveListIsAll);
+        if (previousActiveListIsAll) {
```
Comment (Kennard123661) [inline src/main/java/seedu/address/logic/commands/UndoableCommand.java]: previousActiveIsUncompletedParcelList

File: src/main/java/seedu/address/model/ModelManager.java [hunk 11]
```diff
@@ -35,9 +37,10 @@
 public class ModelManager extends ComponentManager implements Model {
     private static final Logger logger = LogsCenter.getLogger(ModelManager.class);
     private static final Predicate<ReadOnlyParcel> deliveredPredicate = p -> p.getStatus().equals(Status.COMPLETED);
+    private static final Index TAB_ALL_PARCELS = Index.fromZeroBased(0);
```
Comment (Kennard123661) [inline src/main/java/seedu/address/model/ModelManager.java]: image

Please use the variables i created here to get the index lol

Comment (Kennard123661) [reply src/main/java/seedu/address/model/ModelManager.java]: the constants*

File: src/main/java/seedu/address/model/ModelManager.java [hunk 12]
```diff
@@ -231,58 +234,95 @@ public void maintainSorted() {
...
+    }
+
+    /**
+     * Method to retrieve the index of a given parcel in the active list.
+     */
+    private int findIndex(ReadOnlyParcel target) {
+        return getActiveList().indexOf(target);
     }
 
+
```
Comment (Kennard123661) [inline src/main/java/seedu/address/model/ModelManager.java]: remove \n

File: src/main/java/seedu/address/model/ModelManager.java [hunk 13]
```diff
@@ -231,58 +234,95 @@ public void maintainSorted() {
...
+    //@@author
+
```
Comment (Kennard123661) [inline src/main/java/seedu/address/model/ModelManager.java]: remove \n

File: src/main/java/seedu/address/model/parcel/Parcel.java [hunk 14]
```diff
@@ -192,6 +192,12 @@ public int compareTo(Object o) {
         Parcel other = (Parcel) o;
         if (other == this) {
             return 0;
+        } else if (this.getDeliveryDate().compareTo(other.getDeliveryDate()) == 0) {
```
Comment (Kennard123661) [inline src/main/java/seedu/address/model/parcel/Parcel.java]: add comments for your if else statements please, its too many if else, so comments would be good to indicate  how you are doing your comparisons

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java [hunk 15]
```diff
@@ -91,6 +95,19 @@
...
+    public static final DateTimeFormatter DATE_BASIC_FORMATTER = DateTimeFormatter.ofPattern("dd-MM-yyyy");
+    public static final String DATE_TODAY = LocalDate.now().format(DATE_BASIC_FORMATTER);
+    public static final String DATE_YESTERDAY = LocalDate.now().minus(Period.ofDays(1)).format(DATE_BASIC_FORMATTER);
+    public static final String DATE_TOMORROW = LocalDate.now().plus(Period.ofDays(1)).format(DATE_BASIC_FORMATTER);
+
+    public static final String DELIVERY_DATE_DESC_TODAY = " " + PREFIX_DELIVERY_DATE + DATE_TODAY;
+    public static final String DELIVERY_DATE_DESC_YESTERDAY = " " + PREFIX_DELIVERY_DATE + DATE_YESTERDAY;
+    public static final String DELIVERY_DATE_DESC_TOMORROW = " " + PREFIX_DELIVERY_DATE + DATE_TOMORROW;
+
+    public static final Index TAB_COMPLETED_PARCELS = Index.fromZeroBased(1);
```
Comment (Kennard123661) [inline src/test/java/seedu/address/logic/commands/CommandTestUtil.java]: Please use the constants I created previously lol

File: src/test/java/seedu/address/testutil/TypicalParcels.java [hunk 16]
```diff
@@ -131,7 +134,7 @@
             .withTags(VALID_TAG_FROZEN, VALID_TAG_FLAMMABLE).build();
     public static final ReadOnlyParcel CARL = new ParcelBuilder().withTrackingNumber(VALID_TRACKING_NUMBER_CARL)
             .withName(VALID_NAME_CARL).withPhone(VALID_PHONE_CARL).withEmail(VALID_EMAIL_CARL)
-            .withAddress(VALID_ADDRESS_CARL).withStatus(VALID_STATUS_OVERDUE)
```
Comment (Kennard123661) [inline src/test/java/seedu/address/testutil/TypicalParcels.java]: Carl is under the OVERDUE status parcels lol

**Review by Kennard123661 — CHANGES_REQUESTED** [review]

File: src/main/java/seedu/address/model/ModelManager.java [hunk 17]
```diff
@@ -35,9 +39,10 @@
 public class ModelManager extends ComponentManager implements Model {
     private static final Logger logger = LogsCenter.getLogger(ModelManager.class);
     private static final Predicate<ReadOnlyParcel> deliveredPredicate = p -> p.getStatus().equals(Status.COMPLETED);
+    private static final Index TAB_ALL_PARCELS = INDEX_FIRST_TAB;
+    private static final Index TAB_COMPLETED_PARCELS = INDEX_SECOND_TAB;
 
-    private static boolean selected = false;
-    private static Index prevIndex = Index.fromZeroBased(0);
+    private static Index tabIndex = Index.fromOneBased(1);
```
Comment (Kennard123661) [inline src/main/java/seedu/address/model/ModelManager.java]: why is this here? if it meant to be initialized, maybe put it in the constructor?

**Review by Kennard123661 — CHANGES_REQUESTED** [review]

File: src/main/java/seedu/address/model/parcel/Parcel.java (hunk 14 above)
Comment (Kennard123661) [reply src/main/java/seedu/address/model/parcel/Parcel.java]: @fustilio add comments for your comparisons

**Comment by Kennard123661** [comment]
image

@Kowalski985  please ignore this

**Review by Kennard123661 — APPROVED** [review]

**Comment by Kennard123661** [comment]
@Kowalski985 please look through and give your approval if possible

**Review by ktnc985 — APPROVED** [review]
